lib/matplotlib/axes/_base.py

```diff
@@ -2185,6 +2186,10 @@ def grid(self, b=None, which='major', axis='both', **kwargs):
             b = True
         b = _string_to_bool(b)
 
+        # We need to disable both grids before replotting the wanted ones
+        self.xaxis.grid(False)
```

Comment:
I think both of those are the correct behaviour as-is.  In the first case you are saying 'toggle both then toggle x' which I would expect to result in only the 'y' grid to be visible and in the second case you are saying 'turn both on and then turn on x' which should result in both being shown.

I find it very unintuitive that changing something about the 'x' grid would have any effect on the 'y' grid.